libpod/image/image.go

```diff
@@ -149,43 +210,344 @@ func (i *Image) ID() string {
+	}
+	newImage := ir.newFromStorage(img)
+	return newImage, nil
+}
+
+// GetImages retrieves all images present in storage
+// Filters can be provided which will determine which images are included in the
+// output. Multiple filters are handled by ANDing their output, so only images
+// matching all filters are included
+func (ir *ImageRuntime) GetImages(params *libpod.ImageFilterParams, filters ...libpod.ImageFilter) ([]*Image, error) {
+	var imagesFiltered []*storage.Image
+	var newImages []*Image
+
+	images, err := ir.store.Images()
 	if err != nil {
 		return nil, err
 	}
 
-	if decomposedImage.hasRegistry {
-		pullNames = append(pullNames, i.InputName)
-	} else {
-		registries, err := libpod.GetRegistries()
+	for _, img := range images {
+
+		info, err := ir.newFromStorage(&img).GetImageInspectInfo()
 		if err != nil {
 			return nil, err
 		}
-		for _, registry := range registries {
-			decomposedImage.registry = registry
-			pullNames = append(pullNames, decomposedImage.assemble())
+		var names []string
+		if len(img.Names) > 0 {
+			names = img.Names
+		} else {
+			names = append(names, "<none>")
+		}
+		for _, name := range names {
+			include := true
+			if params != nil {
+				params.ImageName = name
+			}
+			for _, filter := range filters {
+				include = include && filter(&img, info)
+			}
+
+			if include {
+				newImage := img
+				newImage.Names = []string{name}
+				imagesFiltered = append(imagesFiltered, &newImage)
+			}
 		}
 	}
-	return pullNames, nil
+
+	for _, i := range imagesFiltered {
+		newImages = append(newImages, ir.newFromStorage(i))
+	}
+	return newImages, nil
 }
 
-// pull is a temporary function for stage1 to be able to pull images during the image
-// resolution tests.  it will be replaced in stage2 with a more robust function.
-func (i *Image) pull(name string, writer io.Writer, r *libpod.Runtime) (string, error) {
-	options := libpod.CopyOptions{
-		Writer:              writer,
-		SignaturePolicyPath: r.GetConfig().SignaturePolicyPath,
+// getImageDigest creates an image object and uses the hex value of the digest as the image ID
+// for parsing the store reference
+func getImageDigest(src types.ImageReference, ctx *types.SystemContext) (string, error) {
+	newImg, err := src.NewImage(ctx)
+	if err != nil {
+		return "", err
+	}
+	defer newImg.Close()
+	digest := newImg.ConfigInfo().Digest
+	if err = digest.Validate(); err != nil {
+		return "", errors.Wrapf(err, "error getting config info")
 	}
-	return i.runtime.PullImage(name, options)
+	return "@" + digest.Hex(), nil
 }
 
-// Remove an image
-// This function is only complete enough for the stage 1 tests.
-func (i *Image) Remove(force bool) error {
-	_, err := i.runtime.RemoveImage(i.image, force)
-	return err
+// TagImage adds a tag to the given image
+func (i *Image) TagImage(tag string) error {
+	tags := i.Names()
+	if StringInSlice(tag, tags) {
+		return nil
+	}
+	tags = append(tags, tag)
+	return i.imageruntime.store.SetNames(i.ID(), tags)
```

Comment:
Do we need to modify the image struct itself during this too?

Reply:
Yeah, nice catch!